module/zfs/dsl_dataset.c

```diff
@@ -291,6 +296,72 @@ dsl_dataset_block_kill(dsl_dataset_t *ds, const blkptr_t *bp, dmu_tx_t *tx,
 	return (used);
 }
 
+struct feature_type_uint64_array_arg {
+	uint64_t length;
+	uint64_t *array;
+};
+
+static void
+unload_zfeature(dsl_dataset_t *ds, spa_feature_t f)
+{
+	switch (spa_feature_table[f].fi_type) {
+	case ZFEATURE_TYPE_BOOLEAN:
+		break;
+	case ZFEATURE_TYPE_UINT64_ARRAY:
+	{
+		struct feature_type_uint64_array_arg *ftuaa = ds->ds_feature[f];
+		kmem_free(ftuaa->array, ftuaa->length * sizeof (uint64_t));
+		kmem_free(ftuaa, sizeof (*ftuaa));
+		break;
+	}
+	default:
+		panic("Invalid zfeature type!");
```

Comment:
If we hit this case, we're hitting a bug in the code. This could be switched to an ASSERT instead of a straight panic, where we just silently ignore the invalid type. Would that be preferable? something like `ASERT3U(pa_feature_table[f].fi_type, <, ZFEATURE_NUM_TYPES);` at the top of the function, and then just return silently or print a warning message in the default case?

Reply:
I'm fine with the `panic()` since this must be a code bug.  It's a little thing, but I do think we should include the `spa_feature_table[f].fi_type` in the message.  If somehow this does happen all we're likely to get in the console log will be this error message.